airflow/providers/amazon/aws/operators/datasync.py

```diff
@@ -124,379 +184,192 @@ def get_hook(self):
-        self.task_arns = hook.get_task_arns_for_location_arns(
-            self.source_location_arns,
-            self.destination_location_arns)
-        self.log.info('Found %s matching TaskArns', len(self.task_arns))
-        return self.task_arns
+        if not self.candidate_source_location_arns:
+            self.log.info("No matching source Locations")
+            return
 
+        if not self.candidate_destination_location_arns:
+            self.log.info("No matching destination Locations")
+            return
 
-class AWSDataSyncUpdateTaskOperator(BaseOperator):
-    """
-    Update an AWS DataSyncTask
-
-    If ``do_xcom_push`` is True, the TaskArns which were updated
-    will be pushed to an XCom.
-
-    :param str aws_conn_id: AWS connection to use.
-    :param str task_arn: The TaskArn to update. If ``None``, the operator will
-        look in xcom_pull for a TaskArn.
-    :param dict update_task_kwargs: The TaskArn will be updated with ``update_task_kwargs``.
-        ``update_task_kwargs`` is used internally like this:
-        ``boto3.update_task(TaskArn=task_arn, **update_task_kwargs)``
-        Example:  ``{'Name': 'xyz', 'Options': ..., 'Excludes': ...}``
-    :raises AirflowException: If ``task_arn`` is None.
-    :raises AirflowException: If ``update_task_kwargs`` is None.
-    """
-    template_fields = ('task_arn',)
-    ui_color = '#44b5e2'
-
-    @apply_defaults
-    def __init__(
-        self,
-        aws_conn_id='aws_default',
-        task_arn=None,
-        update_task_kwargs=None,
-        *args,
-        **kwargs
-    ):
-        super().__init__(*args, **kwargs)
-
-        # Assignments
-        self.aws_conn_id = aws_conn_id
-        self.task_arn = task_arn
-        self.update_task_kwargs = update_task_kwargs
+        self.log.info("Finding DataSync TaskArns that have these LocationArns")
+        self.candidate_task_arns = hook.get_task_arns_for_location_arns(
+            self.candidate_source_location_arns,
+            self.candidate_destination_location_arns,
+        )
+        self.log.info("Found candidate DataSync TaskArns %s",
+                      self.candidate_task_arns)
+
+    def _choose_task_from_list(self, task_arn_list):
+        """Select 1 DataSync TaskArn from a list"""
+        if self.choose_task_callable:
+            return self.choose_task_callable(task_arn_list)
+        if not task_arn_list:
+            return None
+        if len(task_arn_list) == 1:
+            return task_arn_list[0]
+        raise AirflowException("Too many DataSync Tasks to choose from.")
+
+    def _choose_location_from_list(self, location_arn_list):
+        """Select 1 DataSync LocationArn from a list"""
+        if self.choose_location_callable:
+            return self.choose_location_callable(location_arn_list)
+        if not location_arn_list:
+            return None
+        if len(location_arn_list) == 1:
+            return location_arn_list[0]
```

Comment:
Same here - I think callable should be only called when there is a doubt.

Reply:
Fixed, thanks